js/isomer.js

```diff
@@ -67,25 +72,215 @@ Isomer.prototype._translatePoint = function (point) {
+      drawn: 0,
+      shapeName: currentPath.shapeName
+
+    };
+  }
+ this.paths.length = 0;
+
+ // topological sort
+  
+  var drawBefore = [];
+  for (var i = 0 ; i < pathList.length ; i++){
+	drawBefore[i] = [];
+  }
+  for (var i = 0 ; i < pathList.length ; i++){
+    for (var j = 0 ; j < i ; j++){
+	  if(this._hasIntersection(pathList[i].polygon, pathList[j].polygon)){
+	    var cmpPath = pathList[i].path.closerThan(pathList[j].path, observer);
+	    if(cmpPath < 0){
+	      drawBefore[i][drawBefore[i].length] = j;
+	    }
+	    if(cmpPath > 0){
+	      drawBefore[j][drawBefore[j].length] = i;
+	    }
+	  }
+	}
+  }
+  var drawThisTurn = 1;
+  var index = 0;
+  while(drawThisTurn == 1){
+	index++;
+	drawThisTurn = 0;
+	for (var i = 0 ; i < pathList.length ; i++){
+	  if(pathList[i].drawn == 0){
+	    var canDraw = 1;
+		for (var j = 0 ; j < drawBefore[i].length ; j++){
+		  if(pathList[drawBefore[i][j]].drawn == 0){canDraw = 0;}
+		}
+		if(canDraw == 1){
+		   this.add(pathList[i].path, pathList[i].color, true, pathList[i].shapeName);
+		   drawThisTurn = 1;
+		   pathList[i].drawn = 1;
+		}
+	  }
+	}
+  }
+  //purge 
+  //could be done more in a smarter order, that's why drawn is is an element of pathList[] and not a separate array
+  for (var i = 0 ; i < pathList.length ; i++){
+    if(pathList[i].drawn == 0){
+      this.add(pathList[i].path, pathList[i].color, true, pathList[i].shapeName);
     }
   }
 };
 
 
+//+ Jonas Raoni Soares Silva
+//@ http://jsfromhell.com/math/is-point-in-poly [rev. #0]
+//see also http://jsperf.com/ispointinpath-boundary-test-speed/2
+function isPointInPoly(poly, pt){
+    for(var c = false, i = -1, l = poly.length, j = l - 1; ++i < l; j = i)
+        ((poly[i].y <= pt.y && pt.y < poly[j].y) || (poly[j].y <= pt.y && pt.y < poly[i].y))
+        && (pt.x < (poly[j].x - poly[i].x) * (pt.y - poly[i].y) / (poly[j].y - poly[i].y) + poly[i].x)
+        && (c = !c);
+    return c;
+}
+
+
+/**
+ * Does polygonA has intersection with polygonB ?
+ * Naive approach done first : approximate the polygons with a rectangle
+ * Then more complex method : see if edges cross, or one contained in the other
+ */
+Isomer.prototype._hasIntersection = function(pointsA, pointsB) {
+  var i, j;
+  var AminX = pointsA[0].x;
+  var AminY = pointsA[0].y;
+  var AmaxX = AminX;
+  var AmaxY = AminY;
+  var BminX = pointsB[0].x;
+  var BminY = pointsB[0].y;
+  var BmaxX = BminX;
+  var BmaxY = BminY;
+  for(i = 0 ; i < pointsA.length ; i++){
+    AminX = Math.min(AminX, pointsA[i].x);
+    AminY = Math.min(AminY, pointsA[i].y);
+	AmaxX = Math.max(AmaxX, pointsA[i].x);
+    AmaxY = Math.max(AmaxY, pointsA[i].y);
+  }
+  for(i = 0 ; i < pointsB.length ; i++){
+    BminX = Math.min(BminX, pointsB[i].x);
+    BminY = Math.min(BminY, pointsB[i].y);
+	BmaxX = Math.max(BmaxX, pointsB[i].x);
+    BmaxY = Math.max(BmaxY, pointsB[i].y);
+  }
+  
+  if(((AminX <= BminX && BminX <= AmaxX) || (BminX <= AminX && AminX <= BmaxX)) && 
```

Comment:
It covers every case, I was not clear enough. Let me explain in another way.
The hasIntersection function returns true if the 2 given polygons intersect.
It works this way : 

IF the bounding boxes of the polygons do not intesect, return false (this is very fast to compute, and it is a common case)

ELSE check if the edge of one polygon crosses an edge of the other, or if one polygon is inside the other (this takes more time to compute).

So the point is that we do this bounding boxes thing to avoid doing the complex part every time.